src/kbucket/bucket.rs

```diff
@@ -1127,6 +1146,69 @@ pub mod tests {
         assert_eq!(MAX_NODES_PER_BUCKET - 1, bucket.num_disconnected());
     }
 
+    /// No duplicate nodes can be inserted via the apply_pending function.
+    #[test]
+    fn full_bucket_applied_no_duplicates() {
+        // First fill the bucket with disconnected nodes.
+        let mut bucket =
+            KBucket::<NodeId, ()>::new(Duration::from_secs(1), MAX_NODES_PER_BUCKET, None);
+        fill_bucket(&mut bucket, connected_state());
+
+        let first = bucket.iter().next().unwrap().clone();
+
+        let third = bucket.iter().nth(2).unwrap().clone();
+
+        // Set the first connected node as disconnected
+        println!(
+            "Result of updating first node state: {:?}",
+            bucket.update_status(&first.key, ConnectionState::Disconnected, None)
+        );
+
+        // Add a connected pending node.
+        let key = Key::from(NodeId::random());
+        let node = Node {
+            key: key.clone(),
+            value: (),
+            status: connected_state(),
+        };
+
+        // Add a pending node
+        if let InsertResult::Pending { disconnected } = bucket.insert(node.clone()) {
+            assert_eq!(&disconnected, &first.key);
+        } else {
+            panic!()
+        }
+        assert!(bucket.pending().is_some());
+
+        // A misc node gets dropped, because it may not pass a filter when updating its connection
+        // status.
+        bucket.remove(&third.key);
+
+        // The pending nodes status gets updated
+        // Apply pending gets called within kbuckets, so we mimic here.
+        // The pending time hasn't elapsed so nothing should occur.
+        println!("Result of apply pending: {:?}", bucket.apply_pending());
```

Comment:
Could we replace all these `println`s with asserts? It seems the values stay the same between runs:

```
Result of updating first node state: Updated
Result of apply pending: None
Result of inserting node: Inserted
Result of apply pending: None
Result of updating the pending nodes status: NotModified
```